splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java

```diff
@@ -29,78 +29,89 @@
+                    to=_instance=new TimestampOracle(timestampBlockManager,blockSize);
+                }
+            }
+        }
+        return to;
+    }
+
+    private TimestampOracle(TimestampBlockManager timestampBlockManager, int blockSize) throws TimestampIOException {
         this.timestampBlockManager=timestampBlockManager;
         this.blockSize = blockSize * TIMESTAMP_INCREMENT;
-		initialize();
-	}
+        initialize();
+    }
 
     /**
      * Read the current state of the block from the timestampBlockManager
      *
      * @throws TimestampIOException
      */
-	private void initialize() throws TimestampIOException {
-			synchronized(this) {
+    private void initialize() throws TimestampIOException {
+            synchronized(this) {
                 _maxReservedTimestamp = timestampBlockManager.initialize();
-				_timestampCounter.set(_maxReservedTimestamp + TIMESTAMP_INCREMENT);
-			}
-			try {
-				registerJMX();
-			} catch (Exception e) {
+                _timestampCounter.set(_maxReservedTimestamp + TIMESTAMP_INCREMENT);
+            }
+            try {
+                registerJMX();
+            } catch (Exception e) {
                 SpliceLogUtils.error(LOG, "Unable to register Timestamp Generator with JMX. Service will function but metrics will not be available.");
-			}
+            }
     }
 
-	public long getNextTimestamp(boolean refresh, boolean increment) throws TimestampIOException {
-		if (refresh) {
-            _maxReservedTimestamp = timestampBlockManager.initialize();
-			_timestampCounter.set(_maxReservedTimestamp);
-		}
-
-		long nextTS = increment ? _timestampCounter.addAndGet(TIMESTAMP_INCREMENT): _timestampCounter.get();
-		long maxTS = _maxReservedTimestamp; // avoid the double volatile read
-		if (nextTS > maxTS) {
-			reserveNextBlock(maxTS);
-		}
-		_numTimestampsCreated.incrementAndGet(); // JMX metric
-		return nextTS;
-	}
+    public long getNextTimestamp(boolean refresh, boolean increment) throws TimestampIOException {
+        if (refresh) {
+            synchronized (this) {
+                _maxReservedTimestamp = timestampBlockManager.initialize();
+                _timestampCounter.set(_maxReservedTimestamp);
+            }
+        }
+
+        long nextTS = increment ? _timestampCounter.addAndGet(TIMESTAMP_INCREMENT): _timestampCounter.get();
+        long maxTS = _maxReservedTimestamp; // avoid the double volatile read
+        reserveNextBlockIfNecessary(maxTS);
+        _numTimestampsCreated.incrementAndGet(); // JMX metric
+        return nextTS;
+    }
+
+    private void reserveNextBlockIfNecessary(long nextTimestamp) throws TimestampIOException {
+        long maxTS = _maxReservedTimestamp; // avoid the double volatile read
+        if (nextTimestamp > maxTS) {
+            reserveNextBlock(maxTS);
+        }
+        // Check if reserving next block was enough
+        if (nextTimestamp > _maxReservedTimestamp) {
+            reserveNextBlockIfNecessary(nextTimestamp);
```

Comment:
Recursion is not justified. A simple loop is clearer and can be used in the calling method:
`while (nextTimestamp  > _maxReservedTimestamp ) {
    reserveNextBlock(_maxReservedTimestamp);
}
`
I wouldn't worry about "double volatile read": it's not on the hot path and synchronization cost along with ZK read make its extra cost negligible.